### ISSUE TITLE
Deadlock in win_iocp_io_context::sutdown() when owned by thread_pool

### ISSUE THREAD

**Issue by glebov-andrey** [chriskohlhoff/asio#431]
Hi.

There appears to be a bug when shutting down a `win_iocp_io_context` that is owned by a `thread_pool`, causing a deadlock.

Here is the reproducer (the `std::optional`s are for explicit destruction and are totally *optional*):
``` c++
#include <optional>
#include <asio/steady_timer.hpp>
#include <asio/thread_pool.hpp>

int main() {
    auto thread_pool = std::optional<asio::thread_pool>{std::in_place, 1};
    auto timer = std::optional<asio::steady_timer>{std::in_place, *thread_pool}; // creates a win_iocp_io_context
    thread_pool->join();
    timer.reset();
    thread_pool.reset(); // DEADLOCK!
}
```

When `win_iocp_io_context::shutdown()` calls `thread_->join()` https://github.com/chriskohlhoff/asio/blob/62c4488abccce5e62b0526bb3f5390f57ab3712c/asio/include/asio/detail/impl/win_iocp_io_context.ipp#L134 it deadlocks because the thread is blocked in `win_iocp_io_context::do_one` on a call to `GetQueuedCompletionStatus` https://github.com/chriskohlhoff/asio/blob/62c4488abccce5e62b0526bb3f5390f57ab3712c/asio/include/asio/detail/impl/win_iocp_io_context.ipp#L419-L421

The only way to cause `GetQueuedCompletionStatus` to return from the interface seems to be `win_iocp_io_context::stop()`. If I call it directly then that fixes the deadlock:
``` c++
// ...
asio::use_service<asio::detail::win_iocp_io_context>(*thread_pool).stop();
thread_pool.reset(); // FINE
```

The same bug arises when creating anything else that uses a `win_iocp_io_context`, e.g. a socket.
Maybe, there should be a call to `stop()` (or something equivalent) before joining the thread in `shutdown()`?

The is reproducible in at least 1.13.0 and 1.14.1 (https://github.com/chriskohlhoff/asio/commit/0a52abce85ab83b920cc7257d8a2703fe69bc79c).
The OS is Windows 10 1903, compiled with VS 2019 and clang-cl 9.0.

**Comment by helmesjo** [comment]
This is sort of a showstopper. Anyone looking into it?

**Comment by ihsandemir** [comment]
I am having the same problem:

Reproducer:
```
// register a timer
// cancel the timer
pool.stop();
pool.join();
// the pool does out of scope and destructed
```
The destructor hangs at:
```
clientTest_STATIC_32.exe!boost::asio::detail::win_thread::join() Line 46
	at C:\Boost\include\boost-1_72\boost\asio\detail\impl\win_thread.ipp(46)
clientTest_STATIC_32.exe!boost::asio::detail::win_iocp_io_context::shutdown() Line 136
	at C:\Boost\include\boost-1_72\boost\asio\detail\impl\win_iocp_io_context.ipp(136)
clientTest_STATIC_32.exe!boost::asio::detail::service_registry::shutdown_services() Line 44
	at C:\Boost\include\boost-1_72\boost\asio\detail\impl\service_registry.ipp(44)
clientTest_STATIC_32.exe!boost::asio::execution_context::shutdown() Line 42
	at C:\Boost\include\boost-1_72\boost\asio\impl\execution_context.ipp(42)
clientTest_STATIC_32.exe!boost::asio::execution_context::~execution_context() Line 35
	at C:\Boost\include\boost-1_72\boost\asio\impl\execution_context.ipp(35)
clientTest_STATIC_32.exe!boost::asio::thread_pool::~thread_pool() Line 61
	at C:\Boost\include\boost-1_72\boost\asio\impl\thread_pool.ipp(61)

The worker thread:
clientTest_STATIC_32.exe!boost::asio::detail::win_iocp_io_context::do_one(unsigned long msec, boost::system::error_code & ec) Line 420
	at C:\Boost\include\boost-1_72\boost\asio\detail\impl\win_iocp_io_context.ipp(420)
clientTest_STATIC_32.exe!boost::asio::detail::win_iocp_io_context::run(boost::system::error_code & ec) Line 202
	at C:\Boost\include\boost-1_72\boost\asio\detail\impl\win_iocp_io_context.ipp(202)
clientTest_STATIC_32.exe!boost::asio::detail::win_iocp_io_context::thread_function::operator()() Line 48
	at C:\Boost\include\boost-1_72\boost\asio\detail\impl\win_iocp_io_context.ipp(48)
clientTest_STATIC_32.exe!boost::asio::detail::win_thread::func<boost::asio::detail::win_iocp_io_context::thread_function>::run() Line 123
	at C:\Boost\include\boost-1_72\boost\asio\detail\win_thread.hpp(123)
clientTest_STATIC_32.exe!boost::asio::detail::win_thread_function(void * arg) Line 119
	at C:\Boost\include\boost-1_72\boost\asio\detail\impl\win_thread.ipp(119)

```

**Comment by chriskohlhoff** [comment]
Addressed in asio version 1.16.1.
